### ISSUE TITLE
should shellcheck warn *potential* SIGPIPE?

### ISSUE THREAD

**Issue by e-kwsm** [koalaman/shellcheck#3506]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows a potential problem:

```sh
#!/bin/bash
set -eo pipefail
seq 100000 | head
echo unreachable
```

#### Here's what shellcheck currently says:

nothing.

#### Here's what I wanted to see:

```
-:3:14: warning: This may send SIGPIPE and abort execution
```

##### OK commands

- `cat`
- `cut`
- `fold`
- `grep` without `-q` etc
- `od`
- `paste`
- `sha1sum` etc
- `shuf`
- `sort`
- `tac`
- `tail`
- `tr`
- `uniq`

##### commands that *may* send SIGPIPE

- [ ] `grep -q` etc (#665; #3498)
- [ ] `head` (#429) → safe alternatives are `sed -n -e '1,10p'`, `sed -e '11,$d'` or `awk 'NR <= 10'`

Of course SIGPIPE is not sent when the grep patten does not match or piped data is small enough (< 1 MiB).

##### commands that *may* send SIGPIPE (unclear)

- [ ] `nawk`—exit
- [ ] `sed`—q

It is hard to check awk and sed because it depends on the script bodies (and awk implementation):

```sh
seq 100000 |
  awk -f scriptfile
# ^~ This may send SIGPIPE and abort execution, depending on the script
```

and eventually

```sh
seq 100000 |
  unknown-command
# ^~ ShellCheck does not know this command and cannot guarantee that SIGPIPE will not sent

# if ShellCheck knows that the command does not read stdin, SC2216 will be issued
```
